framework/src/base/NonlinearSystemBase.C

```diff
@@ -2308,6 +2314,69 @@ NonlinearSystemBase::residualGhosted()
 }
 
 void
+NonlinearSystemBase::updateJacobianSparsity(SparseMatrix<Number> & jacobian)
+{
+  // This is computed in libMesh according to the mesh
+  const std::vector<dof_id_type> & basic_n_nz = dofMap().get_n_nz();
+  const std::vector<dof_id_type> & basic_n_oz = dofMap().get_n_oz();
+
+  std::vector<dof_id_type> n_nz(basic_n_nz.begin(), basic_n_nz.end());
+  std::vector<dof_id_type> n_oz(basic_n_oz.begin(), basic_n_oz.end());
+
+  _fe_problem.updateGeomSearch();
+
+  std::map<dof_id_type, std::vector<dof_id_type>> graph;
+
+  findImplicitGeometricCouplingEntries(_fe_problem.geomSearchData(), graph);
+
+  if (_fe_problem.getDisplacedProblem())
+    findImplicitGeometricCouplingEntries(_fe_problem.getDisplacedProblem()->geomSearchData(),
+                                         graph);
+
+  const dof_id_type first_dof_on_proc = dofMap().first_dof(processor_id());
+  const dof_id_type end_dof_on_proc = dofMap().end_dof(processor_id());
+
+  // The total number of dofs on and off processor
+  const dof_id_type n_dofs_on_proc = dofMap().n_local_dofs();
+  const dof_id_type n_dofs_not_on_proc = dofMap().n_dofs() - dofMap().n_local_dofs();
+
+  for (const auto & git : graph)
+  {
+    dof_id_type dof = git.first;
+    dof_id_type local_dof = dof - first_dof_on_proc;
+
+    if (dof < first_dof_on_proc || dof >= end_dof_on_proc)
+      continue;
+
+    const std::vector<dof_id_type> & row = git.second;
+
+    for (const auto & coupled_dof : row)
+    {
+      if (coupled_dof < first_dof_on_proc || coupled_dof >= end_dof_on_proc)
+      {
+        if (n_oz[local_dof] < n_dofs_not_on_proc)
+          n_oz[local_dof]++;
+      }
+      else
+      {
+        if (n_nz[local_dof] < n_dofs_on_proc)
+          n_nz[local_dof]++;
+      }
+    }
+  }
+#ifdef LIBMESH_HAVE_PETSC
+#if !PETSC_VERSION_LESS_THAN(3, 7, 2)
+  MatXAIJSetPreallocation(static_cast<PetscMatrix<Number> &>(jacobian).mat(),
+                          dofMap().block_size(),
+                          numeric_petsc_cast(n_nz.empty() ? nullptr : &n_nz[0]),
+                          numeric_petsc_cast(n_oz.empty() ? nullptr : &n_oz[0]),
+                          nullptr,
+                          nullptr);
```

Comment:
@fdkong I am not sure if this was executed during your test problem or not, but with my implementation I am getting following error coming from this line, 
```
[0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[0]PETSC ERROR: Argument out of range
[0]PETSC ERROR: New nonzero at (17,332) caused a malloc
Use MatSetOption(A, MAT_NEW_NONZERO_ALLOCATION_ERR, PETSC_FALSE) to turn off this check
[0]PETSC ERROR: See http://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
[0]PETSC ERROR: Petsc Release Version 3.7.5, Jan, 01, 2017
[0]PETSC ERROR: ../../phase_field-opt on a arch-darwin-c-opt named ECNs-MacBook-Pro.local by sudipta Mon Jul 24 16:19:29 2017
[0]PETSC ERROR: Configure options --prefix=/opt/moose/petsc/mpich_petsc-3.7.5/clang-opt-superlu --download-hypre=1 --with-ssl=0 --with-debugging=no --with-pic=1 --with-shared-libraries=1 --with-cc=mpicc --with-cxx=mpicxx --with-fc=mpif90 --download-fblaslapack=1 --download-metis=1 --download-parmetis=1 --download-superlu_dist=1 --download-mumps=1 --download-scalapack=1 -CC=mpicc -CXX=mpicxx -FC=mpif90 -F77=mpif77 -F90=mpif90 -CFLAGS="-fPIC -fopenmp" -CXXFLAGS="-fPIC -fopenmp" -FFLAGS="-fPIC -fopenmp" -FCFLAGS="-fPIC -fopenmp" -F90FLAGS="-fPIC -fopenmp" -F77FLAGS="-fPIC -fopenmp"
[0]PETSC ERROR: #1 MatSetValues_SeqAIJ() line 485 in /private/var/folders/4r/1yq9mtf16n500m2rvj16td400000gn/T/moose_package_build_temp/petsc-3.7.5-mpich-clang-opt/petsc-3.7.5/src/mat/impls/aij/seq/aij.c
[0]PETSC ERROR: #2 MatSetValues() line 1190 in /private/var/folders/4r/1yq9mtf16n500m2rvj16td400000gn/T/moose_package_build_temp/petsc-3.7.5-mpich-clang-opt/petsc-3.7.5/src/mat/interface/matrix.c
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor
```
I am not sure if I am missing anything. I updated 'n_oz/n_nz', is there anything else I should modify to use this? You can see my implementation here (this is still work in progress) , https://github.com/SudiptaBiswas/moose/commit/f88b4c152aacc5f63ee4a29a197a70c49f0e9737